### ISSUE TITLE
Gradient Accumulation Control

### ISSUE THREAD

**Issue by paradox715** [NVlabs/stylegan2-ada-pytorch#102]
I've noticed that the control of gradient accumulation is a bit challenging - but perhaps I'm not familiar enough with the code.  Is there a bit of guidance on how to adjust the code to increase the amount of accumulation before the weight updates?  

In particular, when running on a card with lower memory at 256x256 it takes me about 1 minute for 1kimg, and takes about 4 minutes per 1kimg when I go to 512x512 (4x makes sense to me due to the scaling of the resolution).  However, because my batch size falls to 1 to accommodate my RAM requirements, I get GAN collapse.  To avoid this, I've successfully reduced the learning rate to about 1/4 of the default, which seems to fix GAN collapse.  Problem is that while I am still processing 1kimg every 4 minutes, I only achieve 1/4 of the weight update, so effectively I'm finding my training time to achieve similar results at the lower resolution has increased by a factor of 16, when it should only have increased by ~4x due to the bigger resolution.  

I would expect if I could just increase the gradient accumulation by 4x, I could keep the higher learning rate and avoid GAN collapse at the same time.  But I'm having a bit of trouble mucking around with this, because the use of batch_gpu and num_gpus in the training_loop.py seems to get overwritten by train.py args, and creates a few other issues when I adjust the code.  

Much appreciated!

**Comment by kampelmuehler** [comment]
have a look at https://github.com/NVlabs/stylegan2-ada-pytorch/blob/d4b2afe9c27e3c305b721bc886d2cb5229458eba/train.py#L190-L191

If you are on a single GPU and set eg.  `args.batch_size = 4` and `args.batch_gpu = 1` it would simulate an effective batch_size of 4 via gradient accumulation here https://github.com/NVlabs/stylegan2-ada-pytorch/blob/d4b2afe9c27e3c305b721bc886d2cb5229458eba/training/training_loop.py#L281-L284 .

Let me know if that works for you!

**Comment by paradox715** [comment]
That makes sense - I think I also need to remove reference to batch size in my the command line

!python train.py --outdir="D:\Models" --data="D:\train.zip" --gpus=1 --snap=3 --metrics=none **--batch=1** --mirror=True --aug=ada --target=0.7 --augpipe=bgcfnc 

or else this next snippet of code will just overwrite the hard-coded 4 & 1 combination you recommended.... I'll give it a shot and see if it works overnight here ...   

https://github.com/NVlabs/stylegan2-ada-pytorch/blob/d4b2afe9c27e3c305b721bc886d2cb5229458eba/train.py#L214-L220

**Comment by paradox715** [comment]
No luck actually - when I run the command line by omitting --batch=1, and manually set the args.batch_size = 4 and args.batch_gpu = 1 in train.py, the setup works ok, but I get an OOM on the first gradient update  :(

**Comment by paradox715** [comment]
Upon a bit more testing, if I tune down the other memory requirements ever so slightly (in my case by reducing the fmaps) it does work.  Thanks a lot! 

It seems like the RAM usage under batch size 1, with gradient accumulation is slightly > than just single batches - but not by a lot (I was just near the edge).

**Comment by kampelmuehler** [comment]
Happy to help :)

**Comment by kampelmuehler** [comment]
... just be aware that all losses that are an average and BatchNorm might not behave as expected.

If you accumulate for `iter_size` iterations, make sure to divide all losses that operate on a mean by `iter_size`.
For BatchNorm there's no easy fix. In extreme cases you can replace with InstanceNorm or GroupNorm.
